\section{Introduction}
\label{sec:introduction}

\subsection{Motivation}
\label{sec:motivation}

Computer information analysis is the only way to work with huge amount of data produced today. In any filed of human work there are data and there are tasks of analysing it. The topic of the Thesis combined two areas of data computing: information visualisation and biological data processing.

Data for processing was provided by Plant Bioinformatics Group of Leibniz Institute of Plant Genetics and Crop Plant Research (IPK), Germany. Data consists of Gene Ontologies and hierarchical clustering, which are are both important tools in biology and medicine to study high-throughput data.

The aim of this work is to provide new visualization approach and implement it to provide useful tool for biologist in their everyday research work.

\subsection{Report structure}
\label{sec:structure}

Section~\ref{sec:introduction} presents the results of the literature review, where been exploring relevant research efforts in the fields of Bioinformatic, Information Visualisation and Bio Visualization, in connection to the work. In Section~\ref{sec:problem} explains the problem and purpose of this work. Section is split in several parts which covers topics: thesis purpose and collaboration, visualisation complexity and goals, description of the input data, data format overview, overview of different graph file formats. In this section, Problem Statement and Goals, have determined the requirements, use cases, and proposed the architecture for thesis application. First part of the Section~\ref{sec:solution} describes visualisation solution for Cluster Ananlisys tree, second part explains Gene Ontology visualisation technique. On of the biggest section in the report is Section~\ref{section:implementation} which covers implementation details, architecture of the system, used libraries, project management tool, etc. Technical details of the visualisation algorithms are explained in the Sections~\ref{sec:cluster} and~\ref{sec:go}. Last part of the report is Section~\ref{sec:conclusion} describes problems that been faced during work and future improvements.

\subsection{Information Visualization}
\label{sec:infovis}

The field of computer-based information visualization draws on ideas from several intellectual traditions: computer science, psychology, semiotics, graphic design, cartography, and art. The two main threads of computer science relevant for visualization are computer graphics and human-computer interaction. The areas of cognitive and perceptual psychology offer important scientific guidance on how humans perceive visual information. A related conceptual framework from the humanities is semiotics, the study of symbols and how they convey meaning. Design, as the name suggests, is about the process of creating artifacts well- suited for their intended purpose. Cartographers have a long history of creating visual representations that are carefully chosen abstractions of the real world. Finally, artists have refined methods for conveying visual meaning in subdisciplines ranging from painting to cinematography.

Information visualization has gradually emerged over the past fifteen years as a distinct field with its own research agenda. The distillation of results from areas with other goals into specific prescriptive advice that can help us design and evaluate visualization systems is nontrivial. Although these traditions have much to offer, effective synthesis of such knowledge into a useful methodology of our own requires arduous gleaning.

The standard argument for visualization is that exploiting visual processing can help people explore or explain data. We have an active field of study because the design challenges are significant and not fully understood. Questions about visual encoding are even more central to information visualization than to scientific visualization. The subfield names grew out of an accident of history, and have some slightly unfortunate con- notations when juxtaposed: information visualization is not unscientific, and scientific visualization is not uninformative. The distinction between the two still not agreed on by all, but the definition used here is that information visualization hinges on finding a spatial mapping of data that is not inherently spatial, whereas scientific visualization uses a spatial layout that is implicit in the data.

\begin{figure}[h!]
\centering
\includegraphics[scale=0.3]{pictures/Tree_graph_example.png}
\caption{Sample visualisation of the tree graph}
\label{fig:tree_graph_example}
\end{figure}

"Graph drawing or Graph layout, as a branch of graph theory, applies topology and geometry to derive two-dimensional representations of graphs. A drawing of a graph is basically a pictorial representation of an embedding of the graph in the plane (generally, with edge intersections allowed), usually aimed at a convenient visualization of certain properties of the graph in question or of the object modeled by the graph. Graph drawing is motivated by applications such as Very-large-scale (VLSI)~\cite{VLSI} integration circuit design , social network analysis, cartography, and bioinformatics, many of which make use of information visualization."~\cite{Graph_drawing}

There are lots of methods and approaches to graph visualization. They are all based on various perception qualities of human. The Radial Dendrograms~\cite{Radial_dendrogram} algorithm is used for results of cluster analysis visualization.

Dendrogram~\cite{Dendrogram} plot is one of the visualisation algorythms for hierarchical structures. It illustrating the outcome of decision tree-type clustering in statistics. Most commonly, dendrograms are drawn in a Cartesian layout, as an upright tree. However, this layout does not make good use of space, it is sparse towards the root and crowded towards the leaf nodes. The spacing between nodes at different levels in the hierarchy is not uniform, which is due to the shrinking number of nodes from bottom to top. For this reason, long, wide-spanning connecting lines are needed to merge nodes at higher levels. A better layout in this respect is the polar or radial layout, where leaf nodes are located on the outer ring and the root is located in the center, as a focal point. A more uniform node spacing results, leading to a better utilization of space and resulting in a better illustration of the class relationships. Recently, Barlow and Neville~\cite{Barlow_Neville} presented an empirical user study for tree layouts (with less than 200 leaves) in which they compare some of the major schemes: organizational chart (a standard drawing of a tree), tree ring (basically a pie chart of circular segments), icicle plot (the cartesian version of the tree ring), and tree map. According to the measured performance within a group of 15 users, the three former methods yielded similar results, with the icicle plot having a slight advantage. However, given the much larger number of leaves in our case (1000 and more) and the fact that the tree ring is the most compact of the three winning configurations, a radial layout seemed to be the most favorable one for our purpose. Radial graph layouts that illustrate hierarchical relationships are very popular, and for the special application of dendrograms, we know only of one other application using a radial layout, the recent one by Kreussler and Schumann~\cite{Kreussler_Schumann} that mentioned before, users often would like to focus on certain portions of the display, while compressing others, without losing context. Fisheye lenses and hyperbolic zooming have been proposed to provide these capabilities. In the context of tree rings Yang, Ward and Rundensteiner~\cite{Yang_Ward} have proposed a system in which users may either perform a polar zoom (i.e. expand the width of one or more adjacent rings while reducing others) or a radial zoom (i.e. expand the arc angle of some adjacent segments while reducing others). Users can perform these operations by pinning down one ring or arc segment and dragging another. A limiting factor here is that users cannot perform both operations simultaneously, which can be awkward in certain instances. To address this shortcoming, our application generalizes these concepts by allowing arbitrary warps of the dendrogram domain, i.e. we allow radial and polar zooms simultaneously.


\subsection{Bioinformatic}
\label{sec:bioinformatic}

"In the last few decades, advances in molecular biology and the equipment available for research in this field have allowed the increasingly rapid sequencing of large portions of the genomes of several species. In fact, to date, several bacterial genomes, as well as those of some simple eukaryotes (e.g., Saccharomyces cerevisiae, or baker's yeast) have been sequenced in full. The Human Genome Project, designed to sequence all 24 of the human chromosomes, is also progressing. Popular sequence databases, such as GenBank and EMBL, have been growing at exponential rates. This deluge of information has necessitated the careful storage, organization and indexing of sequence information. Information science has been applied to biology to produce the field called Bioinformatics.


The simplest tasks used in bioinformatics concern the creation and maintenance of databases of biological information. Nucleic acid sequences (and the protein sequences derived from them) comprise the majority of such databases. While the storage and or organization of millions of nucleotides is far from trivial, designing a database and developing an interface whereby researchers can both access existing information and submit new entries is only the beginning. The most pressing tasks in bioinformatics involve the analysis of sequence information"~\cite{Biology}


Here we can find short introduction and history of Bioinformatics: "Bioinformatics is the application of information technology to the field of molecular biology. The term bioinformatics was coined by Paulien Hogeweg in 1978 for the study of informatic processes in biotic systems. Bioinformatics now entails the creation and advancement of databases, algorithms, computational and statistical techniques, and theory to solve formal and practical problems arising from the management and analysis of biological data. Over the past few decades rapid developments in genomic and other molecular research technologies and developments in information technologies have combined to produce a tremendous amount of information related to molecular biology. It is the name given to these mathematical and computing approaches used to glean understanding of biological processes. Common activities in bioinformatics include mapping and analysing DNA and protein sequences, aligning different DNA and protein sequences to compare them and creating and viewing 3-D models of protein structures."~\cite{Bioinformatic}
The primary goal of bioinformatics is to increase our understanding of biological processes. What sets it apart from other approaches, however, is its focus on developing and applying computationally intensive techniques (e.g., data mining, machine learning algorithms, and visualization) to achieve this goal. Major research efforts in the field include sequence alignment, gene finding, genome assembly, protein structure alignment, protein structure prediction, prediction of gene expression and protein-protein interactions, genome-wide association studies and the modelling of evolution.

\subsection{Gene Ontology}
\label{sec:gene_ontology}

As stated above, there are different knowledge databases for biologic information storage. Gene Ontology~\cite{GO_website} project is one of the first-rate international projects. The Gene Ontology, or GO, is a major bioinformatics initiative to unify the representation of gene and gene product attributes across all species. The aims of the Gene Ontology project are threefold; firstly, to maintain and further develop its controlled vocabulary of gene and gene product attributes; secondly, to annotate genes and gene products, and assimilate and disseminate annotation data; and thirdly, to provide tools to facilitate access to all aspects of the data provided by the Gene Ontology project. The GO is part of a larger classification effort, the Open Biomedical Ontologies (OBO)~\cite{OBO}. The Gene Ontology project provides an ontology of defined terms representing gene product properties. The ontology covers three domains; cellular component, the parts of a cell or its extracellular environment; molecular function, the elemental activities of a gene product at the molecular level, such as binding or catalysis; and biological process, operations or sets of molecular events with a defined beginning and end, pertinent to the functioning of integrated living units: cells, tissues, organs, and organisms. Each GO term within the ontology has a term name, which may be a word or string of words; a unique alphanumeric identifier; a definition with cited sources; and a name space indicating the domain to which it belongs. Terms may also have synonyms, which are classed as being exactly equivalent to the term name, broader, narrower, or related; references to equivalent concepts in other databases; and comments on term meaning or usage. The GO ontology is structured as a directed acyclic graph, and each term has defined relationships to one or more other terms in the same domain, and sometimes to other domains. The GO vocabulary is designed to be species-neutral, and includes terms applicable to prokaryotes and eukaryotes, single and multicellular organisms. The GO ontology is not static, and additions, corrections and alterations are suggested by, and solicited from, members of the research and annotation communities, as well as by those directly involved in the GO project. For example, an annotator may request a specific term to represent a metabolic pathway, or a section of the ontology may be revised with the help of community experts. Suggested edits are reviewed by the ontology editors, and implemented where appropriate.


\subsection{Clustering}
\label{sec:clustering}

"Data clustering (or just clustering), also called cluster analysis, segmentation analysis, taxonomy analysis, or unsupervised classification, is a method of creating groups of objects, or clusters, in such a way that objects in one cluster are very similar and objects in different clusters are quite distinct. Data clustering is often confused with classification, in which objects are assigned to predefined classes. In data clustering, the classes are also to be defined."~\cite{data_clustering_book}


Clustering algorithms can be applied in many fields, for instance:

\begin{itemize}
\item Marketing: finding groups of customers with similar behaviour given a large database of customer data containing their properties and past buying records;
\item Biology: classification of plants and animals given their features;
\item Libraries: book ordering;
\item Insurance: identifying groups of motor insurance policy holders with a high average claim cost; identifying frauds;
\item City-planning: identifying groups of houses according to their house type, value and geographical location;
\item Earthquake studies: clustering observed earthquake epicentres to identify dangerous zones;
\item WWW: document classification; clustering web log data to discover groups of similar access patterns.
\end{itemize}